docs/lang/articles/deployment/tutorial.md

```diff
@@ -4,185 +4,215 @@ sidebar_position: 1
+
+```c++
+#include <taichi/cpp/taichi.hpp>
 ```
 
-### 3. Get Taichi Runtime Library (TiRT)
+Next, create a Taichi runtime with target architecture. We will further load the compiled artifacts from `module.tcm` and load our `paint` kernel from the module.
+
+```C++
+ti::Runtime runtime(TI_ARCH_VULKAN);
+ti::AotModule aot_module = runtime.load_aot_module("module.tcm");
+ti::Kernel kernel_paint = aot_module.get_kernel("paint");
+```
 
+The `paint` kernel acts as a fragment shader, and thus we allocate a piece of memory as its render target. Considering that the kernel accept an Ndarray input argument, we allocate memory through TiRT's `allocate_ndarray` interface. We set both the width and height of the allocated Ndarray memory to 512. Each element of the Ndarray is a 3D vector of 32-bit floating-point numbers, representing the RGB values of the pixel.
 
-![TiRT](../static/assets/runtime.png)
+```c++
+const uint32_t w = 512;
+const uint32_t h = 512;
+ti::NdArray<float> pixels = runtime.allocate_ndarray<float>({w, h}, {3}, true);
+```
 
-With the completion of your Python work, you are now ready to develop your application! The compiled artifacts and TiRT that you have saved are all you require.
+Then, we specify the argument `pixels` for the kernel, where the index `0` indicates the position in the kernel's argument list. Launch the kernel, and wait for the Taichi kernel process to finish.
 
-Currently, TiRT is included with the `taichi-nightly` Python package. Please note that there is not yet a strict version compatibility enforced, so it is strongly advised to use a Python Taichi and TiRT that were built from the same exact commit for compatibility purposes.
+```c++
+kernel_paint[0] = pixels;
+kernel_paint.launch();
+runtime.wait();
+```
 
-TODO: We'll figure out a proper way to release it once the versioning issue is improved.
+Finally, the `pixels` Ndarray holds the kernel output. Before we read the output pixel data, we must map a device memory to a user-addressable space. The image data is saved in a plain text ppm format with a utility function `save_ppm`. For the ppm format, please refer to [Wikipedia](https://en.wikipedia.org/wiki/Netpbm#File_formats).
 
-```bash
-# Install python taichi:
-pip install -i https://pypi.taichi.graphics/simple/ taichi-nightly
-# Get the runtime library:
-pip download --no-deps -i https://pypi.taichi.graphics/simple/ taichi-nightly
-# For example
-unzip taichi_nightly-1.3.0.post20221102-cp38-cp38-manylinux_2_27_x86_64.whl
-export TAICHI_C_API_INSTALL_DIR=$PWD/taichi_nightly-1.3.0.post20221102.data/data/c_api/
+```c++
+auto pixels_data = (const float*)pixels.map();
+save_ppm(pixels_data, w, h, "result.ppm");
+pixels.unmap();
 ```
 
-Currently, only TiRT for Linux systems is included in the nightly distributions. If you need one for Android / Windows, please see the FAQ below to build it from source.
+The complete C++ source code is shown below, which is saved as `app.cpp` in the same directory as `app.py`.
```

Comment:
We might just paste a link to github source file here eventually but we can keep it to help reviewing atm.